browser/skus/skus_sdk_service_factory.cc

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/skus/skus_sdk_service_factory.h"
+
+#include "brave/browser/profiles/profile_util.h"
+#include "chrome/browser/profiles/incognito_helpers.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/storage_partition.h"
+
+// static
+SkusSdkServiceFactory* SkusSdkServiceFactory::GetInstance() {
+  return base::Singleton<SkusSdkServiceFactory>::get();
+}
+
+// static
+SkusSdkService* SkusSdkServiceFactory::GetForContext(
+    content::BrowserContext* context) {
+  return static_cast<SkusSdkService*>(
+      GetInstance()->GetServiceForBrowserContext(context, true));
+}
+
+SkusSdkServiceFactory::SkusSdkServiceFactory()
+    : BrowserContextKeyedServiceFactory(
+          "SkusSdkService",
+          BrowserContextDependencyManager::GetInstance()) {}
+
+SkusSdkServiceFactory::~SkusSdkServiceFactory() = default;
+
+KeyedService* SkusSdkServiceFactory::BuildServiceInstanceFor(
+    content::BrowserContext* context) const {
+  // Skus functionality not supported in private / Tor / guest windows
+  if (!brave::IsRegularProfile(context)) {
```

Comment:
this should go in `GetForContext`, BuildServiceInstanceFor should not return null and we should open issues for the factories that do this incorrectly like news, etc..

Reply:
actually I guess it's ok to return nullptr here and the example below is even more confusing because it does both

```
ApkWebAppService* ApkWebAppServiceFactory::GetForProfile(Profile* profile) {
  // ApkWebAppService is not supported if web apps aren't available.
  if (!web_app::AreWebAppsEnabled(profile))
    return nullptr;

  return static_cast<ApkWebAppService*>(
      GetInstance()->GetServiceForBrowserContext(profile, true));
}

// static
ApkWebAppServiceFactory* ApkWebAppServiceFactory::GetInstance() {
  return base::Singleton<ApkWebAppServiceFactory>::get();
}

ApkWebAppServiceFactory::ApkWebAppServiceFactory()
    : BrowserContextKeyedServiceFactory(
          "ApkWebAppService",
          BrowserContextDependencyManager::GetInstance()) {
  DependsOn(ArcAppListPrefsFactory::GetInstance());
  DependsOn(web_app::WebAppProviderFactory::GetInstance());
}

ApkWebAppServiceFactory::~ApkWebAppServiceFactory() {}

KeyedService* ApkWebAppServiceFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  Profile* profile = static_cast<Profile*>(context);
  if (!arc::IsArcAllowedForProfile(profile))
    return nullptr;

  return new ApkWebAppService(profile);
}
```

Reply:
you can leave this for now while I investigate more, but this should be the only place you check it and everything else should check the service for nullptr

Reply:
on second thought we should be doing this in `GetBrowserContextToUse` as discussed on slack

Reply:
@bridiver could be an old comment now; but if not, I'm confused as what the logic would be inside `GetBrowserContextToUse`. All the examples I'm finding are doing:
```
return chrome::GetBrowserContextRedirectedInIncognito(context);
```
Which we don't want to do, as this isn't supported in Incognito/Tor. As @goodov proposed, I can move the checks to `BuildServiceInstanceFor`

Reply:
Checks moved in https://github.com/brave/brave-core/pull/10847/commits/529737f5e820c19972f937a4fdfb296afe572f8d